pkg/geo/s2.go

```diff
@@ -89,26 +89,96 @@ func loopAreaKm2(loop *s2.Loop) float64 {
 	return (loop.Area() * earthAreaKm2) / (4.0 * math.Pi)
 }
 
-// Covering calculates the S2 covering of a set of S2 points. Will try the loop
-// in both clockwise and counter clockwise.
+// chordSegmentsIntersect determines if two chord segments (segment 1 from p1a
+// to p1b and segment 2 from p2a to p2b) on a sphere intersect.
+func chordSegmentsIntersect(p1a s2.Point, p1b s2.Point, p2a s2.Point, p2b s2.Point) bool {
+	// Normal of plane containing great circle connecting p1a to p1b
+	n1 := p1a.Cross(p1b.Vector)
+
+	// Normal of plane containing great circle connecting p1a to p1b
+	n2 := p2a.Cross(p2b.Vector)
+
+	// Possible chord intersection point (other one is ip.Mul(-1))
+	ip := n1.Cross(n2)
+
+	// Chord segments can't intersect if they're both on the same side of the
+	// great circle planar intersection points
+	rp1a := p1a.Cross(ip)
+	rp1b := p1b.Cross(ip)
+	if rp1a.Dot(rp1b) > 0 {
+		return false
+	}
+
+	rp2a := p2a.Cross(ip)
+	rp2b := p2b.Cross(ip)
+	if rp2a.Dot(rp2b) > 0 {
+		return false
+	}
+
+	// Chord segments only intersect if they both have their shortest segments
+	// along their great circle intersect the same possible intersection point
+
+	// Length of chord connecting p1a-ip-p1b
+	l1p := p1a.Angle(ip) + ip.Angle(p1b.Vector)
+
+	// Length of chord connecting p2a-ip-p2b
+	l2p := p2a.Angle(ip) + ip.Angle(p2b.Vector)
+
+	if l1p.Radians() <= math.Pi && l2p.Radians() <= math.Pi {
+		return true
+	} else if l1p.Radians() >= math.Pi && l2p.Radians() >= math.Pi {
+		return true
+	}
+	return false
+}
+
+// validateLoop returns an error if any of the edges formed by the specified
+// points intersect each other.  There is an edge between the last and first
+// vertices.
+func validateLoop(points []s2.Point) error {
```

Comment:
My apologies if this is not the same thing, but do you think we can use this? https://github.com/golang/geo/blob/e86565bf3f350d722923af04ce21a09e28d4b133/s2/loop.go#L225

Reply:
We can and do call that method [via loop.Validate](https://github.com/interuss/dss/pull/446/files#diff-0838de1360740955856c6f0c3d004c22R162), however it [does not check for edge intersections](https://github.com/golang/geo/blob/e86565bf3f350d722923af04ce21a09e28d4b133/s2/loop.go#L245).  Once [this TODO](https://github.com/golang/geo/blob/e86565bf3f350d722923af04ce21a09e28d4b133/s2/loop.go#L215) is resolved, we may be able to switch back to using only the s2 library to validate a polygon instead of checking for edge intersections ourselves.